modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java

```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
      */
     @Override
     public final HybridTimestamp update(HybridTimestamp requestTime) {
-        while (true) {
-            long now = currentTime();
+        long requestTimeLong = requestTime.longValue();
 
+        while (true) {
             // Read the latest time after accessing UTC time to reduce contention.
-            long oldLatestTime = latestTime;
+            long oldLatestTime = this.latestTime;
 
-            long newLatestTime = max(requestTime.longValue() + 1, max(now, oldLatestTime + 1));
+            if (oldLatestTime >= requestTimeLong) {
+                return hybridTimestamp(LATEST_TIME.incrementAndGet(this));
+            }
 
-            // TODO https://issues.apache.org/jira/browse/IGNITE-23707 avoid CAS on logical part update.
+            long now = currentTime();
+
+            if (now > requestTimeLong) {
+                if (LATEST_TIME.compareAndSet(this, oldLatestTime, now)) {
+                    return hybridTimestamp(now);
```

Comment:
Why don't we notify listeners in these cases? Clock is still moved forward

Reply:
Yes, but here the clock is not changed by jumping up. It means it is not faster than the scheduler in clock waiter.

Reply:
What do you mean by "not changed by jumping up".
Clock here is moved up to current physical time.

Reply:
Yes, it happens the same as in the now() invocation, where we do not notify the clock waiter.

Reply:
After thinking a bit, It seems there should be no problem here